### ISSUE TITLE
deletion of a block leaves dangling yellow boxes + strange behavior with auto-revert

### ISSUE THREAD

**Issue by rdiaz02** [sheijk/hideshowvis#10]
Suppose we have two different emacs sessions that have opened the same file, and we have auto-revert enabled (and hideshowvis, of course). If we make edits on one frame, funny things can start to accumulate on the other emacs frame.
Note: I am using the code with the fix in https://github.com/sheijk/hideshowvis/issues/8#issuecomment-1972120920.

In the example below, I will be using this file

```
(defun f0 ()
  (setq a 0))

(defun f1 ()
  (setq a 1))

(defun f1 ()
  (setq a 1))

(defun f2 ()
  (setq a 2))

(defun f3 ()
  (setq a 3
	))

(defun f4 ()
  (setq a 4
	))
;; open this file in two separate emacs instances
;; have auto-revert enabled
;; then remove from lines 1 to 7. Observe what happens in the other instance
;; If you continue adding and removing in this emacs session
;; without M-x revert-buffer in the other, you start piling up
;; lots of yellow thingies.
```

Steps to reproduce
================

1. emacs -Q
2. load the following code. Notice the global-auto-revert AND the hs-hide-all in the emacs-lisp-mode-hook
```
(global-auto-revert-mode 1)
(add-hook 'emacs-lisp-mode-hook 'hs-minor-mode)
(add-hook 'emacs-lisp-mode-hook 'hs-hide-all)
(add-to-list 'load-path "~/Sources/emacs-packages/hideshowvis/")
(require 'hideshowvis)
(add-hook 'emacs-lisp-mode-hook 'hideshowvis-enable)
(add-hook 'emacs-lisp-mode-hook 'hideshowvis-symbols)

```

3. Now, some screenshots of two emacs frames from two independent emacs instances that show the file above:

Right after we open the two files
![i1](https://github.com/sheijk/hideshowvis/assets/1688611/c344f623-04a1-4282-b3c7-a91575dc82f5)

On the right, we delete the two "f1". This is already looking funny: the yellow boxes in the place of the no-longer existing "f1"
![i2-right-after-rm](https://github.com/sheijk/hideshowvis/assets/1688611/c6390ba3-fe72-4e5f-b9af-0cf9ac6c9d51)

On the right, we do show all. 
![i3-hs-show-all](https://github.com/sheijk/hideshowvis/assets/1688611/87a880d4-136b-429c-9266-72dbf614468f)

We save on the right. Notice what happened on the left (because of the auto-revert)
![i4-right-after-saving-notice-left](https://github.com/sheijk/hideshowvis/assets/1688611/f7041a91-d6d7-48ba-b1e1-8b2bf53e94d8)

We can make things OK if we intervene manually: do `M-x revert-buffer` on the left.
![i5-after-revert-buffer-on-right](https://github.com/sheijk/hideshowvis/assets/1688611/b1e7f45c-f639-4112-b343-7cbfaceb5c2b)

Now, on the right we removed the top. Notice the lonely (and actually in place of no code) box with the (2) on the top left of  the left frame.
![i6-after-rm-top](https://github.com/sheijk/hideshowvis/assets/1688611/8e12f29b-7d0d-4bbd-a6f6-ab259e08507c)

Having done a bunch of repeated additions and deletions on the right frame, we can easily end up with a left frame like this
![i76-after-several-rm-and-undo](https://github.com/sheijk/hideshowvis/assets/1688611/4c14d210-ac5e-4c0e-8bbd-321f4bcada3a)

With very large files, an accumulation of these operations I think leads Emacs to eventually start spinning with 100% CPU usage.

I think the main issue is that removing a block can lead to having yellow boxes left dangling; auto-revert composes the problem, because you can accummulate those.

I have no idea if the first issue is solvable. The second I solved as follows:

```
;; This won't work: I think the other hooks interfere and leave this without effect
;; (add-hook 'global-auto-revert-mode-hook 'hs-show-all)

(defun my/before-revert-show-all (&optional ignore-auto noconfirm preserve-modes)
  "hs-show-all if in `prog-mode` before reverting."
  (when (derived-mode-p 'prog-mode)
    (hs-show-all)))

;; This solves the problem
(advice-add 'revert-buffer :before #'my/before-revert-show-all)

;; Optional: hide the blocks again
(defun my/after-revert-hide-all (&optional ignore-auto noconfirm preserve-modes)
  "hs-hide-all if in `prog-mode` before reverting."
  (when (derived-mode-p 'prog-mode)
    (hs-hide-all)))

(advice-add 'revert-buffer :after #'my/after-revert-hide-all)
```

Now, this is probably a corner case (hideshowvis + hs-hide-all + global-auto-revert in two or more different emacs). So maybe not worth looking into, given the solution. I submit this, in case someone else stumbles upon this.

**Comment by sheijk** [comment]
I can reproduce the error. I've tried to remove the after-string properties in `before-revert-hook` and re-adding them in `after-revert-hook`. But after removing lines 1-7 there are now two overlays at char 9-9 (in the middle of the remaining "f1") which cause the marker string to show up there twice plus all the valid overlays following after that. I haven't figured out how hs updates it's overlays nor found a way to get them fixed, yet

Code:
```elisp
(defun hideshowvis-remove-overlays ()
  (message "hideshowvis-remove-overlays")
  ;; (remove-overlays (point-min) (point-max) 'hideshowvis-hs t)
  ;; (remove-overlays (point-min) (point-max) 'hs t)
  ;; (remove-overlays)
  (dolist (ov (overlays-in (point-min) (point-max)))
    (unless (null (overlay-get ov 'hs))
      (overlay-put ov 'after-string nil)
      ;; (hideshowvis-display-code-line-counts ov)
      ))
  )

(defun hideshowvis-update-all-overlays ()
  (message "hideshowvis-update-all-overlays")
  (dolist (ov (overlays-in (point-min) (point-max)))
    (unless (null (overlay-get ov 'hs))
      ;; (overlay-put ov 'after-string nil)
      (hideshowvis-display-code-line-counts ov)
      ))
  )

(add-hook 'before-revert-hook 'hideshowvis-remove-overlays)
(add-hook 'after-revert-hook 'hideshowvis-update-all-overlays)
```
plus `(message "hideshowvis-display-code-line-counts %s, %s" (pp-to-string ov) (pp-to-string (overlay-properties ov)))` to dump the overlays

**Comment by sheijk** [comment]
Simply deleting hs overlays which are only one character wide works. Need to think whether that's a good idea, though

```elisp
;; add into hideshowvis-update-all-overlays
      (if (equal (overlay-start ov) (overlay-end ov))
          (delete-overlay ov)
        (hideshowvis-display-code-line-counts ov))
```

**Comment by sheijk** [comment]
Fixed by c2ab91bd77810e4f083179fdfd3839953b03c450
